### PR TITLE
feat: Carry Forward Leave Expiry

### PR DESCRIPTION
**Pull request by Mangesh-Khairnar**
- Add feature to expire carry forwarded leaves for particular leave type

Screenshot 2019-05-30 at 10 31 22 PM

- Add Leave Ledger which maintains all the leaves related data.

Screenshot 2019-05-15 at 11 42 07 PM

- Made reason field in leave application mandatory
Screenshot 2019-05-30 at 10 29 02 PM

- Change positions of several fields in leave application
Screenshot 2019-05-30 at 10 29 16 PM

- Change position of field "is active" in leave period
 
Screenshot 2019-05-30 at 10 26 13 PM

### CODE REVIEW

File: erpnext/hr/doctype/leave_allocation/leave_allocation.py [hunk 1]
```diff
@@ -87,29 +91,69 @@ def validate_back_dated_allocation(self):
...
-			self.from_date, self.to_date)
-
-		if flt(leaves_taken) > flt(self.total_leaves_allocated):
-			if frappe.db.get_value("Leave Type", self.leave_type, "allow_negative"):
-				frappe.msgprint(_("Note: Total allocated leaves {0} shouldn't be less than already approved leaves {1} for the period").format(self.total_leaves_allocated, leaves_taken))
-			else:
-				frappe.throw(_("Total allocated leaves {0} cannot be less than already approved leaves {1} for the period").format(self.total_leaves_allocated, leaves_taken), LessAllocationError)
+	def create_leave_ledger_entry(self, submit=True):
+		if self.unused_leaves:
+			expiry_days = frappe.db.get_value("Leave Type", self.leave_type, "expire_carried_forward_leaves")
```
Comment (netchampfaris) [inline erpnext/hr/doctype/leave_allocation/leave_allocation.py]: `expire_carry_forwarded_leaves_after_days`

Check point number 5 of https://github.com/frappe/erpnext/wiki/Naming-Guidelines#field-naming

File: erpnext/hr/doctype/leave_application/leave_application.js [hunk 2]
```diff
@@ -49,7 +49,7 @@ frappe.ui.form.on("Leave Application", {
 				async: false,
 				args: {
 					employee: frm.doc.employee,
-					date: frm.doc.posting_date
+					date: frm.doc.from_date? frm.doc.from_date:frm.doc.posting_date
```
Comment (netchampfaris) [inline erpnext/hr/doctype/leave_application/leave_application.js]:
```suggestion
					date: frm.doc.from_date || frm.doc.posting_date
```

File: erpnext/hr/doctype/leave_type/leave_type.json [hunk 3]
```diff
@@ -355,10 +379,81 @@
...
+   "allow_bulk_edit": 0,
+   "allow_in_quick_entry": 0,
+   "allow_on_submit": 0,
+   "bold": 0,
+   "collapsible": 0,
+   "columns": 0,
+   "depends_on": "",
+   "description": "Calculated in days",
+   "fetch_if_empty": 0,
+   "fieldname": "expire_carried_forward_leaves",
```
Comment (netchampfaris) [inline erpnext/hr/doctype/leave_type/leave_type.json]: `expire_carry_forwarded_leaves` sounds better and correct

File: erpnext/hr/doctype/leave_type/leave_type.json [hunk 4]
```diff
@@ -355,10 +379,81 @@
...
+   "fieldtype": "Int",
+   "hidden": 0,
+   "ignore_user_permissions": 0,
+   "ignore_xss_filter": 0,
+   "in_filter": 0,
+   "in_global_search": 0,
+   "in_list_view": 0,
+   "in_standard_filter": 0,
+   "label": "Expire Carried Forward Leaves",
```
Comment (netchampfaris) [inline erpnext/hr/doctype/leave_type/leave_type.json]: `Expire Carry Forwarded Leaves (Days)` and remove description

File: erpnext/hr/doctype/leave_type/leave_type.py [hunk 5]
```diff
@@ -2,9 +2,16 @@
...
+from frappe import _
 
 from frappe.model.document import Document
 
 class LeaveType(Document):
-	pass
\ No newline at end of file
+	def validate(self):
+		if self.is_lwp:
+			leave_allocation = frappe.db.sql_list("""select name from `tabLeave Allocation` where leave_type=%s""", (self.name))
```
Comment (netchampfaris) [inline erpnext/hr/doctype/leave_type/leave_type.py]: Use ORM

File: erpnext/hr/doctype/leave_type/leave_type.py [hunk 6]
```diff
@@ -2,9 +2,16 @@
...
+			if leave_allocation:
+				frappe.throw(_("""Leave application is linked with leave allocations {0}. Leave application cannot be set as leave without pay""").format(", ".join(leave_allocation))) #nosec
```
Comment (netchampfaris) [inline erpnext/hr/doctype/leave_type/leave_type.py]:
```suggestion
				frappe.throw(_("Leave application is linked with leave allocations {0}. Leave application cannot be set as leave without pay").format(", ".join(leave_allocation))) #nosec
```

File: erpnext/hr/utils.py [hunk 7]
```diff
@@ -263,38 +263,63 @@ def get_leave_period(from_date, to_date, company):
 	if leave_period:
 		return leave_period
 
+def generate_leave_encashment():
+	''' Generates a draft leave encashment on allocation expiry '''
+	from erpnext.hr.doctype.leave_encashment.leave_encashment import create_leave_encashment
+	if frappe.db.get_single_value('HR Settings', 'auto_leave_encashment'):
+		leave_type = frappe.db.sql_list("SELECT name FROM `tabLeave Type` WHERE `allow_encashment`=1")
```
Comment (netchampfaris) [inline erpnext/hr/utils.py]: ORM please

File: erpnext/hr/utils.py [hunk 8]
```diff
@@ -263,38 +263,63 @@ def get_leave_period(from_date, to_date, company):
...
-				between from_date and to_date and docstatus=1 and leave_type='{1}'"""
-				.format(today, e_leave_type.name), as_dict=1)
-			for allocation in leave_allocations:
-				leave_policy = get_employee_leave_policy(allocation.employee)
-				if not leave_policy:
+
+	for e_leave_type in e_leave_types:
+		leave_allocations = frappe.db.sql("""select name, employee, from_date, to_date from `tabLeave Allocation` where '{0}'
+			between from_date and to_date and docstatus=1 and leave_type='{1}'"""
+			.format(today, e_leave_type.name), as_dict=1)
```
Comment (netchampfaris) [inline erpnext/hr/utils.py]: Avoid `.format`, use `%s` for auto escaping

File: erpnext/hr/utils.py [hunk 9]
```diff
@@ -263,38 +263,63 @@ def get_leave_period(from_date, to_date, company):
...
+		for allocation in leave_allocations:
+			leave_policy = get_employee_leave_policy(allocation.employee)
+			if not leave_policy:
+				continue
+			if not e_leave_type.earned_leave_frequency == "Monthly":
+				if not check_frequency_hit(allocation.from_date, today, e_leave_type.earned_leave_frequency):
+					continue
+			annual_allocation = frappe.db.sql("""select annual_allocation from `tabLeave Policy Detail`
```
Comment (netchampfaris) [inline erpnext/hr/utils.py]: ORM

File: erpnext/patches.txt [hunk 10]
```diff
@@ -603,6 +603,7 @@ erpnext.patches.v11_1.set_salary_details_submittable
 erpnext.patches.v11_1.rename_depends_on_lwp
 execute:frappe.delete_doc("Report", "Inactive Items")
 erpnext.patches.v11_1.delete_scheduling_tool
+execute:frappe.delete_doc_if_exists("Page", "support-analytics")
```
Comment (netchampfaris) [inline erpnext/patches.txt]: ?

File: erpnext/patches/v12_0/generate_leave_ledger_entries.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+	allocation_list = get_allocation_records()
+
+	for allocation in allocation_list:
+		if not frappe.db.exists("Leave Ledger Entry", {'transaction_type': 'Leave Allocation', 'transaction_name': allocation.name, 'is_expired': 1}):
+			allocation.update(dict(doctype="Leave Allocation"))
+			allocation_obj = frappe.get_doc(allocation)
+			expire_allocation(allocation_obj)
+
+def get_allocation_records():
+	return frappe.db.sql("""
```
Comment (netchampfaris) [inline erpnext/patches/v12_0/generate_leave_ledger_entries.py]: ORM

File: erpnext/patches/v12_0/generate_leave_ledger_entries.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+			to_date,
+			carry_forward
+		FROM `tabLeave Allocation`
+		WHERE
+			docstatus=1
+		ORDER BY to_date ASC
+	""", as_dict=1)
+
+def get_leaves_application_records():
+	return frappe.db.sql("""
```
Comment (netchampfaris) [inline erpnext/patches/v12_0/generate_leave_ledger_entries.py]: ORM

File: erpnext/patches/v12_0/generate_leave_ledger_entries.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+				total_leave_days,
+				from_date,
+				to_date
+			FROM `tabLeave Application`
+			WHERE
+				docstatus=1
+		""", as_dict=1)
+
+def get_leave_encashment_records():
+	return frappe.db.sql("""
```
Comment (netchampfaris) [inline erpnext/patches/v12_0/generate_leave_ledger_entries.py]: ORM

**Review by netchampfaris — CHANGES_REQUESTED** [review]
Use raw queries only when they provide atleast 3 times the performance from ORM

**Review by netchampfaris — DISMISSED** [review]

File: erpnext/hr/doctype/leave_type/leave_type.py [hunk 14]
```diff
@@ -5,13 +5,20 @@
...
 	def validate(self):
 		if self.is_lwp:
-			leave_allocation = frappe.db.sql_list("""select name from `tabLeave Allocation` where leave_type=%s""", (self.name))
+			leave_allocation = frappe.get_all("Leave Allocation", filters={
+				'leave_type': self.name,
+				'from_date': ("<=", today()),
+				'to_date': (">=", today())
+			}, ['name'])
+			leave_allocation = [l['name'] for l in leave_allocation]
+			frappe.db("""select name from `tabLeave Allocation` where leave_type=%s""", (self.name))
```
Comment (netchampfaris) [inline erpnext/hr/doctype/leave_type/leave_type.py]: ??

**Review by netchampfaris — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/leave_type/leave_type.py (hunk 14 above)
Comment (Mangesh-Khairnar) [reply erpnext/hr/doctype/leave_type/leave_type.py]: It checks if the user is trying to change leave type when a leave allocation is active. If the leave allocation is not currently active, the user can change the leave type.

File: erpnext/hr/doctype/leave_type/leave_type.py [hunk 15]
```diff
@@ -2,9 +2,23 @@
...
\ No newline at end of file
+	def validate(self):
+		if self.is_lwp:
+			leave_allocation = frappe.get_all("Leave Allocation", filters={
+				'leave_type': self.name,
+				'from_date': ("<=", today()),
+				'to_date': (">=", today())
+			}, fields=['name'])
+			leave_allocation = [l['name'] for l in leave_allocation]
+			frappe.db("""select name from `tabLeave Allocation` where leave_type=%s""", (self.name))
```
Comment (netchampfaris) [inline erpnext/hr/doctype/leave_type/leave_type.py]: `frappe.db("""select name from `tabLeave Allocation` where leave_type=%s""", (self.name))
`

Shouldn't it be `frappe.db.sql`, or should it be removed?

**Review by netchampfaris — CHANGES_REQUESTED** [review]

**Review by netchampfaris — APPROVED** [review]

File: erpnext/hr/doctype/leave_type/leave_type.py (hunk 15 above)
Comment (Mangesh-Khairnar) [reply erpnext/hr/doctype/leave_type/leave_type.py]: added changes

**Pull request merged**